python_modules/libraries/dagster-k8s/dagster_k8s/container_context.py

```diff
@@ -204,6 +215,218 @@ def get_environment_dict(self) -> Mapping[str, str]:
+                    f"Attempted to create a pod with fields that violated the allowed list: {', '.join(disallowed_fields)}"
+                )
+
+        validated_container_context = self
+
+        if only_allow_user_defined_env_vars is not None:
+            validated_container_context = self._filter_user_defined_env_vars(
+                set(only_allow_user_defined_env_vars)
+            )
+
+        return validated_container_context
+
+    def _filter_user_defined_k8s_config_env_vars(
+        self,
+        user_defined_k8s_config: UserDefinedDagsterK8sConfig,
+        only_allow_user_defined_env_vars: Set[str],
+        discarded_env_var_names: Set[str],
+    ) -> UserDefinedDagsterK8sConfig:
+        """Filters out any env vars from the supplied UserDefinedDagsterK8sConfig
+        that are not in the supplied set of env var names and adds the names of
+        any env vars that were discarded to the passed-in discarded_env_var_names set.
+        """
+        if "env" not in user_defined_k8s_config.container_config:
+            return user_defined_k8s_config
+
+        filtered_env = []
+        for env_dict in user_defined_k8s_config.container_config["env"]:
+            env_key = env_dict["name"]
+            if env_key in only_allow_user_defined_env_vars:
+                filtered_env.append(env_dict)
+            else:
+                discarded_env_var_names.add(env_key)
+
+        return UserDefinedDagsterK8sConfig.from_dict(
+            {
+                **user_defined_k8s_config.to_dict(),
+                "container_config": {
+                    **user_defined_k8s_config.container_config,
+                    "env": filtered_env,
+                },
+            }
+        )
+
+    def _filter_user_defined_env_vars(
+        self,
+        only_allow_user_defined_env_vars: Set[str],
+    ) -> "K8sContainerContext":
+        discarded_env_var_names = set()
+
+        new_run_k8s_config = self._filter_user_defined_k8s_config_env_vars(
+            self.run_k8s_config,
+            only_allow_user_defined_env_vars,
+            discarded_env_var_names,
+        )
+
+        new_server_k8s_config = self._filter_user_defined_k8s_config_env_vars(
+            self.server_k8s_config,
+            only_allow_user_defined_env_vars,
+            discarded_env_var_names,
+        )
+
+        new_env_vars = []
+        for env_var in self.env_vars:
+            env_var_name = get_env_var_name(env_var)
+            if env_var_name in only_allow_user_defined_env_vars:
+                new_env_vars.append(env_var)
+            else:
+                discarded_env_var_names.add(env_var_name)
+
+        new_env = []
+        for env_map in self.env:
+            if env_map["name"] in only_allow_user_defined_env_vars:
+                new_env.append(env_map)
+            else:
+                discarded_env_var_names.add(env_map["name"])
+
+        if discarded_env_var_names:
+            logging.warning(
```

Comment:
a hard failure would be consistent with the higher level rule set you picked, and might be nicer to debug

Reply:
agreed that would be more consistent - the main motivation here is for places (like the dagster cloud control plane) where we set environment variables programatically and want to be able to add new ones without breaking clients - bit of discussion on that in a comment earlier on the PR